File: spring/src/test/java/cardsproject/controller/cards/CardAbilityControllerTest.java
package cardsproject.controller.cards;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.transaction.annotation.Transactional;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
@Transactional
public class CardAbilityControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void list_returns200() throws Exception {
        mockMvc.perform(get("/api/card_abilities"))
            .andExpect(status().isOk());
    }

    @Test
    void create_returns201() throws Exception {
        mockMvc.perform(post("/api/card_abilities")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"abilityText\": \"test\" }"))
            .andExpect(status().isCreated());
    }

    @Test
    void show_returns200or404() throws Exception {
        mockMvc.perform(get("/api/card_abilities/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 200 || status == 404;
            });
    }

    @Test
    void delete_returns204or404() throws Exception {
        mockMvc.perform(delete("/api/card_abilities/1"))
            .andExpect(result -> {
                int status = result.getResponse().getStatus();
                assert status == 204 || status == 404;
            });
    }
    @Test
    void create_fails_when_keyword_ability_requires_keyword_violated() throws Exception {
        // Keyword ability must have a keyword name: antecedent true, consequent missing → 400
        mockMvc.perform(post("/api/card_abilities")
            .contentType(MediaType.APPLICATION_JSON)
            .content("{ \"abilityText\": \"test\", \"abilityType\": \"KEYWORD\", \"keyword\": null }"))
            .andExpect(status().isBadRequest());
    }
}

File: phoenix/test/cards/card_ability_controller_test.exs
defmodule CardsProjectWeb.Cards.CardAbilityControllerTest do
  use ExUnit.Case, async: false
  use CardsProjectWeb.ConnCase

  @valid_params %{
    "ability_text" => "test",
    "ability_type" => "Keyword"
  }

  describe "GET /api/card_abilities" do
    test "returns 200 with list", %{conn: conn} do
      conn = get(conn, "/api/card_abilities")
      assert json_response(conn, 200) |> is_list()
    end
  end

  describe "POST /api/card_abilities" do
    test "creates record and returns 201", %{conn: conn} do
      conn = post(conn, "/api/card_abilities", @valid_params)
      assert %{"id" => _id} = json_response(conn, 201)
    end
  end

  describe "GET /api/card_abilities/:id" do
    test "returns 200 for existing record", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_ability(@valid_params)
      conn = get(conn, "/api/card_abilities/#{record.id}")
      assert json_response(conn, 200)
    end
  end

  describe "PUT /api/card_abilities/:id" do
    test "updates and returns 200", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_ability(@valid_params)
      conn = put(conn, "/api/card_abilities/#{record.id}", @valid_params)
      assert json_response(conn, 200)
    end
  end

  describe "DELETE /api/card_abilities/:id" do
    test "deletes and returns 204", %{conn: conn} do
      {:ok, record} = CardsProject.Cards.create_card_ability(@valid_params)
      conn = delete(conn, "/api/card_abilities/#{record.id}")
      assert response(conn, 204)
    end
  end
end
